doc/extensibility/dataflow_best_practices.md

```diff
@@ -0,0 +1,112 @@
+# DataFlow Best Coding Practices
+These are suggestions to help avoid some common pitfalls when using DataFlow blocks.
+
+## Use Slim DataFlow blocks 
+You should always use the Slim version blocks provided by CPS, unless you need advanced capabilities that are not supported (like processing multiple works at the same time). See [DataFlow Examples](dataflow_example.md) for more information.
+
+## Expose `BroadcastBlock` via API
+You should typically expose `BroadcastBlock` via API instead of TransformBlock, etc
+  - Other blocks will retain data when there is no subscriber
```

Comment:
> Other blocks will retain data when there is no subscriber

Is this true?

```c#
ActionBlock<int> ab = new(Console.WriteLine);

BroadcastBlock<int> broadcast = new(cloningFunction: null);

broadcast.Post(1);
broadcast.Post(2);
broadcast.Post(3);

broadcast.LinkTo(ab);
```

Even though the subscriber is linked late, values 1, 2 and 3 are all printed to the console.

Reply:
This from a Word doc on Dataflow helps clear this up:

> Unlike `BufferBlock<T>`, `BroadcastBlock<T>` doesn’t hold on to data unnecessarily.  After a particular datum has been offered to all targets, that element will be overwritten by whatever piece of data is next in line.

So it will buffer when no target is linked to it, but won't once one is. In practice I suppose we're likely to have linked targets. By default these buffers are unbounded, so probably worth calling out that having propagator blocks that don't link to anywhere can result in large buffer allocations if they receive input.

However that doc also says:

> If no targets are linked from a `BroadcastBlock<T>`, the data will effectively be dropped.

...which doesn't match my experience above.

Reply:
Aaaaand now I cannot repro the above code producing 1, 2 and 3. It only spits out the last value. I must have made a mistake.

Reply:
So, in that case, do you recommend making any changes to the initial statement?

Reply:
No. Though now that I think of it some code samples such as the one I include here might be helpful to demonstrate the difference you're highlighting.